### PR TITLE
T-45 - Create custom exceptions

### PR DESCRIPTION
**Pull request by AntonBessarabov**
Implemented custom exceptions

### CODE REVIEW

**Comment by AntonBessarabov** [comment]
[T-45 Create custom exceptions](https://trello.com/c/6Nh6HhPl/45-t-45-create-custom-exceptions)

File: PortfolioKeeper/CustomExceptions/BusinessException.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+{
+    [Serializable]
+    public class BusinessException : CustomException
+    {
+        public BusinessException()
+        {
+        }
+
+        public BusinessException(
+            string message,
```
Comment (DashaDrahun) [inline PortfolioKeeper/CustomExceptions/BusinessException.cs]: Move this to the previous line, please

File: PortfolioKeeper/Controllers/AssetController.cs [hunk 2]
```diff
@@ -150,7 +143,7 @@ public string Valuations(int id, string time)
 
             if (userId == null)
             {
-                throw new SecurityException("User claim value not found");
+                throw new ValidationException("User claim value not found", null, "AssetController");
```
Comment (stas-litvinenka) [inline PortfolioKeeper/Controllers/AssetController.cs]: no need to specify controller name

File: PortfolioKeeper/Controllers/AssetController.cs [hunk 3]
```diff
@@ -76,60 +77,52 @@ public ActionResult SelectAssetPropertiesTypes(byte parentTypeId, byte childrenT
...
         {
-            if (ModelState.IsValid)
-            {
-                assetService.AddConcreteAsset(car, (byte) AssetTypeValue.Car, User.Identity.GetUserId());
-            }
-            else
+            if (!ModelState.IsValid)
             {
                 return View(car);
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/Controllers/AssetController.cs]: remove line

File: PortfolioKeeper/BusinessLogic/Services/SecurityCheckService.cs [hunk 4]
```diff
@@ -55,15 +57,15 @@ public void CheckAssetAccess(int assetId, string userId)
...
-            if (asset.Portfolio.AppUser.Id == userId)
-                return;
-
-            logger.Warn($"No permissions. SecurityCheckService.CheckAssetAccess assetId: {assetId} userId:{userId}");
+            if (asset.Portfolio.AppUser.Id != userId)
+            {
+                logger.Warn($"No permissions. SecurityCheckService.CheckAssetAccess assetId: {assetId} userId:{userId}");
 
-            throw new SecurityException($"No permission to access the asset portfolioId:{assetId}");
+                throw new ValidationException($"No permission to access the asset ({assetId}, {userId})");
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/SecurityCheckService.cs]: more informative - No permission to access the asset. AssetId={assetId}, UserId={userId}. try to use one logging style

File: PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs [hunk 5]
```diff
@@ -20,30 +21,32 @@ public class AssetValuationService : IValuationService<AssetValuationDto, Asset>
...
-                              ?? throw new ArgumentNullException(nameof(calculator));
+                ?? throw new ArgumentNullException(nameof(calculator));
             this.valuationHelper = valuationHelper
-                              ?? throw new ArgumentNullException(nameof(valuationHelper));
-            this.securityService = securityService ??
-                                      throw new ArgumentNullException(nameof(securityService));
+                ?? throw new ArgumentNullException(nameof(valuationHelper));
+            this.securityService = securityService
+                ?? throw new ArgumentNullException(nameof(securityService));
+
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Services/AssetValuationService.cs]: remove line

File: PortfolioKeeper/BusinessLogic/Helpers/ValuationHelper.cs [hunk 6]
```diff
@@ -1,22 +1,29 @@
...
             if (price.IsNullOrWhiteSpace())
+            {
                 throw new ArgumentNullException(nameof(price));
+            }
 
             decimal valueNumber;
-            var parseValue = Decimal.TryParse(price, NumberStyles.Number,
-                CultureInfo.InvariantCulture, out valueNumber);
+            var parseValue = Decimal.TryParse(
+                price,
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Helpers/ValuationHelper.cs]: no need to put each value on new line here

**Pull request merged**